### PR TITLE
feat: add support for composing data and errors in graphql

### PR DESCRIPTION
**Pull request by tdeekens**
This pull request implements one suggestion (without recommending it specifically discussed here https://github.com/mswjs/msw/issues/401.

It thereby implements  allowing `json` to use a `json({}, { merge: true })` option which `data` and `errors` of the `graphql` context can use. Internally it parses the prior `res.body` and shallowly merges the new `BodyType` in.

This internally is used by `data` and `errors` to support composing both with each other.

### CODE REVIEW

File: src/context/json.test.ts
Comment (tdeekens) [inline src/context/json.test.ts]: Just a test to verify that we don't deep merge things.

File: src/context/errors.test.ts
Comment (tdeekens) [inline src/context/errors.test.ts]: This is what we're after in the end. I tested it both in the `data` and `errors` context as both support it and do strictly are not concerned how it is achieved underneath them.

File: src/context/json.ts
Comment (tdeekens) [inline src/context/json.ts]: I just find this easier to read when having defaults on top a function.

File: src/context/json.ts
Comment (tdeekens) [inline src/context/json.ts]: So that if we allow more options in the future and one is `false` we don't overwrite it with others.

File: src/context/json.ts
Comment (tdeekens) [inline src/context/json.ts]: I think `BodyType` is typed wrong internally in the library from what I can tell. The `BodyType` is the generic of what is passed in as an e.g. object. It's then put as a JSON string on the `response`. The response typing however states that `res.body` is a `BodyType` while it is a stringified version of it. That's why there was the `as any` and we now need the `as string` to my understanding.

It is also why CI fails. TypeScript would like us to convert to an `unknown` first. Should be, if the assumption is correct, typing issue too?

File: src/context/data.ts
Comment (tdeekens) [inline src/context/data.ts:10]: Not sure why this wasn't using `json` just as `errors`. A likely worthwhile change regardless.

File: src/context/data.test.ts
Comment (tdeekens) [inline src/context/data.test.ts]: What won't work with shallow merging in `json` here is

```js
res(
  data({ foo: 'bar'}),
  data({ bar: 'bar'})
)
```

as it arrives at `json` as `data` which is one level nested meaning here we would only get `{ bar: 'bar }` has that's how `json` merges. Just for awareness.

**Comment by tdeekens** [comment]
Any updates here. Should I be doing something different? How can I help to get this out there if we need it :)

File: src/context/data.ts
Comment (kettanaito) [reply src/context/data.ts:10]: Most likely for historical reasons. I agree with your refactor here: it should reuse the `json` utility.

File: src/context/data.test.ts
Comment (kettanaito) [reply src/context/data.test.ts]: That is incorrect behavior. We should merge _deeply_, with the latest `data` call taking priority. 

```js
res(data({ a: 1 }), data({ b: 2, c: { d: 3 } }))
// { data: { a: 1, b: 2, c: { d: 3 } } }
```

Using the existing `mergeRight` utility already does the deep merging. Let's reflect that in test suites.

File: src/context/data.test.ts
Comment (kettanaito) [inline src/context/data.test.ts]: Need to add a test when using multiple `data()` utilities and assert that they merge the given payloads into one.

File: src/context/errors.test.ts
Comment (kettanaito) [inline src/context/errors.test.ts]: Need to also add a test where we use multiple `errors()` utilities and assert that they produce a list of merged errors.

File: src/context/json.ts
Comment (kettanaito) [reply src/context/json.ts]: This is perhaps an optimization done too early. Let's utilize the default argument values for now, and if we get to supporting more options we'd adopt the approach you suggested.

```js
const json = (body, options = { merge: false }) => {}
```

File: src/context/json.ts
Comment (kettanaito) [inline src/context/json.ts]: Seeing how tedious it becomes to operate with a JSON body I think we should drop its stringification within the `json` utility and instead do so on the `res()` chain itself. Then we won't have to parse and stringify the body back and forth, making the code simpler.

```js
const json = (body, options) => {
  return (res) => {
    res.headers.set('Content-Type', 'application/json')
    // No need to parse or stringify
    res.body = options.merge ? mergeRight(res.body, body) : body
    return res
  }
}
```

**Review by kettanaito — CHANGES_REQUESTED** [review]
Hey, @tdeekens! Thank you for preparing these changes. I think they look awesome.

I've left a few comments, could you please take a look and address those? Everything is open to discussion, as always. Looking forward to seeing this feature land :)

File: src/context/data.test.ts
Comment (tdeekens) [reply src/context/data.test.ts]: Thanks. Will add a spec.

File: src/context/errors.test.ts
Comment (tdeekens) [reply src/context/errors.test.ts]: Thanks. Will add a spec.

File: src/context/json.ts
Comment (tdeekens) [reply src/context/json.ts]: That sounds nice @kettanaito can you maybe point me where the `res() chain itself` resides and give me a hint please. I struggle to find my way around.

File: src/context/json.ts
Comment (kettanaito) [reply src/context/json.ts]: I think you're right, the type may be wrong. Does it work with `res.body` typed as `string`?

File: src/context/json.ts
Comment (kettanaito) [reply src/context/json.ts]: Of course. Here's where the `res` is defined:

https://github.com/mswjs/msw/blob/2b133436cd405a9ebf6d6882d5b8e47997b943f6/src/response.ts#L40-L56

The `resolvedResponse` is the final instance of response composed by it and sent to the worker. I think we can stringify the response body before the return statement. Just pay attention if this is a safe thing to do for all `Content-Type` of responses (i.e. should we stringify body for `text/plain` response?).

Comment (tdeekens) [reply src/context/json.ts]: Got ya, cheers. I think we should start with just `JSON.stringifying` the JSON Content-type. This will also solve the issue of `json` returning a `string` which caused some typing annoyances (e.g. the old `any` type). Will look into it soon.

Comment (kettanaito) [reply src/context/json.ts]: Sounds reasonable. Keep in mind there are multiple JSON content-types, we may check them as something like `res.headers.get('content-type')?.endsWith('json')`.

Comment (tdeekens) [reply src/context/json.ts]: Here would be my first propsal: 7bf236d. I will comment also inline.

File: src/context/json.ts
Comment (tdeekens) [inline src/context/json.ts:18]: This removes any typing quirk I'd hope.

File: src/context/json.ts
Comment (tdeekens) [inline src/context/json.ts]: The `|| {}` is only for any first merge to avoid an `if`.

File: src/response.ts
Comment (tdeekens) [inline src/response.ts]: Would also be a tranform I guess we just compose in. Let me know what you think.

File: src/context/errors.ts
Comment (kettanaito) [inline src/context/errors.ts]: We need to be careful in dropping the `ResponseTransformer` return type here. It may cause some unexpected type issues when using `ctx.errors` in combination with other utils.

File: src/context/json.test.ts
Comment (kettanaito) [inline src/context/json.test.ts]: I think we can limit this to 2 bodies in total. The fact that given JSON gets merged deeply is asserted in the unit tests for the `mergeRight` utility, there's no need to verify that once more. In `json` utility test we just test that given `merge: true` it merges 2 bodies, if not—it doesn't.

File: src/context/json.ts
Comment (kettanaito) [reply src/context/json.ts:18]: Can you please explain more on how this works? I see that you've renamed 

```diff
- BodyType
+ BodyTypeJSON
```

But I'm certainly missing something here...

File: src/context/json.ts
Comment (kettanaito) [inline src/context/json.ts]: This `if` is unnecessary: you can assign a ternary expression to the `res.body`

```js
res.body = merge ? mergeRight(...) : body
```

That produces the same output as the if/else statement.

File: src/context/errors.ts
Comment (tdeekens) [reply src/context/errors.ts]: Ok. Thanks.

File: src/context/json.ts
Comment (tdeekens) [reply src/context/json.ts]: True.

File: src/response.ts
Comment (kettanaito) [inline src/response.ts]: Let's drop the encapsulation of a single transformer into an array :) I know the intention here, but on practice this will remain an array with a single member forever. We can apply `JSONBodyResponseTransformer` directly.

File: src/response.ts
Comment (kettanaito) [inline src/response.ts]: That's an interesting design decision we are to make: should we stringify JSON body as a part of the response composition chain (potentially allowing to de-stringify it after you called `res()`), or should we do that right before sending the mocked response to the worker? 🤔

**Review by kettanaito — CHANGES_REQUESTED** [review]
I've left a few comments, yet it looks awesome! We are close to merging this.

**Comment by kettanaito** [comment]
Could you please check your local build? There seem to be a type violation (per CI):

```
[!] (plugin rpt2) Error: /home/circleci/release/src/context/json.ts(23,7): semantic error TS2322: Type 'Record<string, any>' is not assignable to type 'BodyTypeJSON'.
  'BodyTypeJSON' could be instantiated with an arbitrary type which could be unrelated to 'Record<string, any>'.
src/context/json.ts
Error: /home/circleci/release/src/context/json.ts(23,7): semantic error TS2322: Type 'Record<string, any>' is not assignable to type 'BodyTypeJSON'.
  'BodyTypeJSON' could be instantiated with an arbitrary type which could be unrelated to 'Record<string, any>'.
    at error (/home/circleci/release/node_modules/rollup/dist/shared/rollup.js:5211:30)
    at throwPluginError (/home/circleci/release/node_modules/rollup/dist/shared/rollup.js:17799:12)
    at Object.error (/home/circleci/release/node_modules/rollup/dist/shared/rollup.js:18418:24)
    at Object.error (/home/circleci/release/node_modules/rollup/dist/shared/rollup.js:17972:38)
    at RollupContext.error (/home/circleci/release/node_modules/rollup-plugin-typescript2/src/rollupcontext.ts:37:18)
    at /home/circleci/release/node_modules/rollup-plugin-typescript2/src/print-diagnostics.ts:41:11
    at arrayEach (/home/circleci/release/node_modules/rollup-plugin-typescript2/node_modules/lodash/lodash.js:516:11)
    at Function._.each [as forEach] (/home/circleci/release/node_modules/rollup-plugin-typescript2/node_modules/lodash/lodash.js:9368:14)
    at printDiagnostics (/home/circleci/release/node_modules/rollup-plugin-typescript2/src/print-diagnostics.ts:9:2)
    at Object.transform (/home/circleci/release/node_modules/rollup-plugin-typescript2/src/index.ts:242:5)
```

**Comment by kettanaito** [comment]
If you run `yarn test:unit` there is a failing test:

```
Summary of all failing tests
 FAIL  src/context/errors.test.ts
  ● Test suite failed to run

    src/context/json.ts:23:7 - error TS2322: Type 'Record<string, any>' is not assignable to type 'BodyTypeJSON'.
      'BodyTypeJSON' could be instantiated with an arbitrary type which could be unrelated to 'Record<string, any>'.

    23       res.body = mergeRight(body, res.body || {})
```

You can also reproduce this by running that test in isolation locally: `yarn test:unit src/context/errors.test.ts`.

File: src/response.ts
Comment (tdeekens) [reply src/response.ts]: Yeah, I was torn here knowing there is a premature abstraction lurking on the corner :). I will simplify it. We can always abstract it later.

File: src/response.ts
Comment (tdeekens) [reply src/response.ts]: Yeah fair point. Initially I just wrapped the call with the transform than passed it down. Let me show what I had initally.

File: src/response.ts
Comment (tdeekens) [inline src/response.ts:61]: Is this what you had in mind? I find it confusing that the `createResponseComposition` returns a `ResponseFunction`. While `compose(...t)(response)` to my understanding doesn't return a `ResponseFunction` but just the `response` which I see also is being passed to the `Object.assign` below?

**Comment by tdeekens** [comment]
@kettanaito thanks for the help. I am stuck on a last typing error and I fail to follow how it's meant to be.

The `resolvedResponse` [here](https://github.com/mswjs/msw/pull/403/commits/a25611cac7928a531a7e08f6ef544b63dba8f982#diff-635146f32166c4a4f9ab579dae7e2253L51) is a `Partial<MockedResponse>` which is combined with the `overrides` which are also a `Partial<MockedResponse>`. If this we have `transformers` we pipe the `resolvedResponse` through them if not we return it [here](https://github.com/mswjs/msw/pull/403/commits/a25611cac7928a531a7e08f6ef544b63dba8f982#diff-635146f32166c4a4f9ab579dae7e2253L68). The `createResponseComposition` itself returns a `ResponseFunction` which returns a `MockedResponse<BodyType>` **not** a `Partial` of it. I am confused how the typing ever worked. For one I assume the `compose` returned an `any` which is assignable to anything. Now the `JSONBodyResponseTransformer` returns a `Partial<MockedReponse>` which seems logical. However, then the return of `createResponseComposition` is not as expected. How do you see this? Can you shed some light here? To me it seems, if I am not mistaken, the last bit to get this through 😄.

**Comment by tdeekens** [comment]
@kettanaito sorry to ping and bother but any feedback here?

**Comment by kettanaito** [comment]
Hey, @tdeekens. Thanks once more for preparing these changes! I think they look fantastic. 

I took this opportunity to refine the pieces around request/response JSON transformation in general, as well as to fix a wrongly typed test suite. 

What concerns your typing question, I've decided to go with the type casting:

```ts
mergeRight(...) as BodyTypeJSON
```

Nevertheless, I'm open to suggestions here. Other than that let's wait for the CI and merge this, if it's green.

File: src/context/data.test.ts
Comment (tdeekens) [inline src/context/data.test.ts:59]: That makes sense. Thought of that too but didn't dare to make that change.

**Review by kettanaito — APPROVED** [review]
The changes looks great. Thank you, @tdeekens, and welcome to contributors! 🎉

**Comment by tdeekens** [comment]
@kettanaito any plans to release this? 🤔

**Comment by kettanaito** [comment]
Hi, @tdeekens. The next version's release took some time, but I think it'll happen this week.

**Pull request merged**
